File: ios/Classes/src/NaverCircleController.swift
//
//  NaverCircleController.swift
//  naver_map_plugin
//
//  Created by Maximilian on 2020/08/31.
//

import NMapsMap
import Flutter

class NCircleController: NSObject{
    let circle: NMFCircleOverlay
    let id: String
    
    init(json: NSDictionary) {
        assert(json["overlayId"] != nil
            && json["center"] != nil
            && json["radius"] != nil)
        circle = NMFCircleOverlay()
        self.id = json["overlayId"] as! String
        
        super.init()
        circle.userInfo = ["circle" : self]
        
        interprete(json: json)
    }
    
    func interprete(json: NSDictionary){
        if let centerData = json["center"] {
            circle.center = toLatLng(json: centerData)
        }
        if let radius = json["radius"] as? Double {
            circle.radius = radius
        }
        if let color = json["color"] as? NSNumber {
            circle.fillColor = toColor(colorNumber: color)
        }
        if let outlineColor = json["outlineColor"] as? NSNumber {
            circle.outlineColor = toColor(colorNumber: outlineColor)
        }
        if let outlineWidth = json["outlineWidth"] as? Double {
            circle.outlineWidth = outlineWidth
        }
        if let zIndex = json["zIndex"] as? Int {
            circle.zIndex = zIndex
        }
        if let globalZIndex = json["globalZIndex"] as? Int {
            circle.globalZIndex = globalZIndex
        }
        if let minZoom = json["minZoom"] as? Double {
            circle.minZoom = minZoom
        }
        if let maxZoom = json["maxZoom"] as? Double {
            circle.maxZoom = maxZoom
        }
    }
    
    func setMap(_ naverMap: NMFNaverMapView?){
        circle.mapView = naverMap?.mapView
    }
}

class NaverCircleController: NSObject{
    private var idToController = Dictionary<String, NCircleController>()
    let naverMap: NMFNaverMapView
    let touchHandler: (NMFOverlay)->Bool
    
    init(naverMap: NMFNaverMapView, touchHandler: @escaping (NMFOverlay)->Bool) {
        self.naverMap = naverMap
        self.touchHandler = touchHandler
    }
    
    func add(jsonArray: Array<Any>) {
        jsonArray.forEach { (json) in
            DispatchQueue.main.async {
                if let data = json as? NSDictionary {
                    let circleToAdd = NCircleController(json: data)
                    circleToAdd.setMap(self.naverMap)
                    circleToAdd.circle.touchHandler = self.touchHandler
                    self.idToController[circleToAdd.id] = circleToAdd
                }
            }
        }
    }
    
    func remove(jsonArray: Array<Any>) {
        jsonArray.forEach { (rawId) in
            DispatchQueue.main.async {
                if let id = rawId as? String, let circleToRemove = self.idToController[id] {
                    circleToRemove.circle.touchHandler = nil
                    circleToRemove.setMap(nil)
                    self.idToController.removeValue(forKey: id)
                }
            }
        }
    }
    
    func modify(jsonArray: Array<Any>) {
        jsonArray.forEach { (json) in
            DispatchQueue.main.async {
                if let data = json as? NSDictionary {
                    let id = data["overlayId"] as! String
                    self.idToController[id]?.interprete(json: data)
                }
            }
        }
    }
    
}

File: android/src/main/kotlin/map/naver/plugin/net/note11/naver_map_plugin/NaverCircleController.kt
package map.naver.plugin.net.note11.naver_map_plugin

import android.os.Handler
import map.naver.plugin.net.note11.naver_map_plugin.Convert.toLatLng
import map.naver.plugin.net.note11.naver_map_plugin.Convert.toColorInt
import com.naver.maps.map.NaverMap
import com.naver.maps.map.overlay.Overlay
import android.os.Looper
import com.naver.maps.map.overlay.CircleOverlay
import java.util.HashMap
import java.util.concurrent.Executors
import kotlin.math.roundToInt

class NaverCircleController(
    private val naverMap: NaverMap,
    private val listener: Overlay.OnClickListener,
    private val density: Float
) {
    private val idToController = HashMap<String?, CircleController?>()
    private val handler = Handler(Looper.getMainLooper())
    fun add(jsonArray: List<Any?>?) {
        if (jsonArray == null || jsonArray.isEmpty()) return
        val service = Executors.newCachedThreadPool()
        service.execute {
            for (json in jsonArray) {
                val data = json as HashMap<String, Any>
                val circle = CircleController(data)
                circle.setOnClickListener(listener)
                idToController[circle.id] = circle
            }
            handler.post {
                val circles: List<CircleController?> = idToController.values.toList()
                for (circle in circles) {
                    circle?.setMap(naverMap)
                }
            }
        }
        service.shutdown()
    }

    fun modify(jsonArray: List<Any?>?) {
        if (jsonArray == null || jsonArray.isEmpty()) return
        for (json in jsonArray) {
            val data = json as HashMap<String, Any>
            val id = data["overlayId"] as String?
            idToController[id]!!.interpret(data)
        }
    }

    fun remove(jsonArray: List<Any?>?) {
        if (jsonArray == null || jsonArray.isEmpty()) return
        for (json in jsonArray) {
            val id = json as String?
            val marker = idToController[id]
            marker?.setOnClickListener(null)
            marker?.setMap(null)
            idToController.remove(id)
        }
    }

    inner class CircleController(json: HashMap<String, Any>) {
        val id: String = json["overlayId"] as String
        private val circle: CircleOverlay = CircleOverlay()

        init {
            circle.tag = this
            interpret(json)
        }

        fun interpret(json: HashMap<String, Any>) {
            if (json.containsKey("center")) circle.center = json["center"]!!.toLatLng()
            if (json.containsKey("radius")) circle.radius = json["radius"] as Double
            if (json.containsKey("color")) circle.color =
                toColorInt(json["color"])
            if (json.containsKey("outlineColor")) circle.outlineColor =
                toColorInt(json["outlineColor"])
            if (json.containsKey("outlineWidth")) circle.outlineWidth =
                (json["outlineWidth"] as Int * density).roundToInt()
            if (json.containsKey("zIndex")) circle.zIndex = json["zIndex"] as Int
            if (json.containsKey("globalZIndex")) circle.globalZIndex =
                json["globalZIndex"] as Int + 1
            if (json.containsKey("minZoom")) circle.minZoom = (json["minZoom"] as Int).toDouble()
            if (json.containsKey("maxZoom")) circle.maxZoom = (json["maxZoom"] as Int).toDouble()
        }

        fun setMap(naverMap: NaverMap?) {
            circle.map = naverMap
        }

        fun setOnClickListener(onClickListener: Overlay.OnClickListener?) {
            circle.onClickListener = onClickListener
        }
    }
}